### ISSUE TITLE
Props Behavior

### ISSUE THREAD

**Issue by webberwang** [codelab-ai/codelab.ai#132]
```
type PropType = 'eval' | 'leaf' | 'single'
  __type: PropType | Array<PropType>,
  value: string
}
```

1. `eval` should only evaluate, it is not a renderProp by default

Currently our `eval` prop is by default a `leaf` render prop, which should not be the behavior. Discussed here. https://github.com/codelab-ai/codelab.ai/issues/108

2. to pass `eval` 1 level, we need to do `['eval', 'single']`
3. to pass `eval` to `leaf`, we need to do `['eval', 'leaf']`
4. When accessing renderProps, the `.value` is omitted
5. All renderProps are accessible on the child on `this.props` property (renderProps don't bind to children automatically, it is job of children to bind the prop, we are emulating behavior in React itself.)
6. `this.props` should also contain props that is passed by the component (not just renderProps, again this would be how React works)

At each level, merge all renderProps coming from the components themselves with the user specified renderProps, with user specified taking merge precedence (meaning if same name user props will overwrite component props). Then pass the merged props to the children component, & should be accessible via `this` in the children `eval` functions.

Instead of `this.onClick.value()`, we should just be able to call `this.onClick()`. Maybe we can add a getter to the object? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

This is discussed here https://github.com/codelab-ai/codelab.ai/issues/113

7. Keep `ctx`, the behavior of `ctx` is that it strips out it's own key. The behavior of eval & passing down still needs to be be specified

We can achieve `ctx` like behavior using a renderProp, but we'll be forced to access the key of the renderProp. Instead, we just want to access `this`. `ctx` just strips out it's own key (it doesn't have `eval` or `leaf` by default anymore)

To put it another way, if `ctx` wasn't special, we would access like this `this.ctx.someProperty` in the children. But we are making `ctx` a special property, which can be accessed like `this.someProperty` instead.

8. Remove Provider for now (comment out, & change examples to use React.Fragment, its behavior is too confusing. We can use React.Fragment to specify render props

The idea is to have small composable units of behavior that we can combine together to create new behavior. This is actually inline with the decorator pattern, which uses chaining to compose different pieces of behavior together. 

https://sourcemaking.com/design_patterns/decorator

You don't have to use this pattern, just something to think about

**Comment by VienNguyen24hdev** [comment]
via Hubstaff
User: Nguyen Luong Vien

Project: Codelab - https://app.hubstaff.com/projects/1093588
Date Range: 09/22/20 - 09/28/20
Work session total: 25:40:04
Billable: Yes

Grand total: 25:40:04
